rxjava2/src/main/java/io/vertx/reactivex/core/json/ObservableUnmarshaller.java

```diff
@@ -0,0 +1,78 @@
+package io.vertx.reactivex.core.json;
+
+import com.fasterxml.jackson.core.type.TypeReference;
+import io.reactivex.ObservableOperator;
+import io.reactivex.Observer;
+import io.reactivex.annotations.NonNull;
+import io.reactivex.disposables.Disposable;
+import io.vertx.core.buffer.Buffer;
+
+import java.io.IOException;
+
+import static io.vertx.core.json.Json.mapper;
+import static java.util.Objects.nonNull;
+
+/**
+ * An operator to unmarshall json to pojos.
+ *
+ * @author <a href="mailto:julien@julienviet.com">Julien Viet</a>
+ */
+public class ObservableUnmarshaller<T, B> implements ObservableOperator<T, B> {
```

Comment:
No it wouldn't, this isn't how you implement a transformer. Here's a modified version:

```java
public abstract class UnmarshallerTransformer<B, T> implements Observable.Transformer<B, T> {

  private final Func1<Buffer, T> converter;

  public UnmarshallerTransformer(Class<T> mappedType) {
    converter = buffer -> {
      try {
        return mapper.readValue(buffer.getBytes(), mappedType);
      } catch (IOException e) {
        throw new RuntimeException(e);
      }
    };
  }

  public UnmarshallerTransformer(TypeReference<T> mappedTypeRef) {
    converter = buffer -> {
      try {
        return mapper.readValue(buffer.getBytes(), mappedTypeRef);
      } catch (IOException e) {
        throw new RuntimeException(e);
      }
    };
  }

  public abstract Buffer unwrap(B buffer);

  @Override
  public Observable<T> call(Observable<B> tObservable) {
    return tObservable.collect(Buffer::buffer, (buffer, b) -> buffer.appendBuffer(unwrap(b)))
      .filter(buffer -> buffer.length() > 0)
      .map(converter)
      .defaultIfEmpty(null);
  }
}
```

It's concise, readable and keeps RxJava internals away.

You use it like this:

```java
  public static <T> Observable.Transformer<Buffer, T> observableTransformer(Class<T> mappedType) {
    return new UnmarshallerTransformer<Buffer, T>(mappedType) {
      @Override
      public Buffer unwrap(Buffer buffer) {
        return buffer;
      }
    };
  }

  Observable<MyPojo> obs = bufObs.compose(observableTransformer(MyPojo.class));
```

The difference is explained in the `#compose` and `#lift` Javadoc.

> If your operator is designed to transform the source Observable as a whole (for instance, by applying a particular set of existing RxJava operators to it) use {@code compose}.